docs/source/miscellaneous/doc/adr/0016-use-tickets-encoding-for-transactions.md

```diff
@@ -499,7 +499,131 @@ the worker pool has a finite size and even with the full batching algorithm in t
+schema versions. This is a range map, as we want to support rollbacks easily. The map also is guaranteed to span the
+ranges [1, +∞) and be connected; note that this does not mean future behaviour is fixed, since in practice this map
+is read together with a validity bound. Thus, even though the map will contain as a key some range [C, +∞) for a
+constant C, it should be valid up to some point V > C - and behaviour at timestamps after V may subsequently be changed.
+
+We attempt to read the latest version of the range map, and if our timestamp falls within the validity bound, we
+retrieve the version. If it does not, we submit an identity transformation, which extends the validity bound of the
+current value. We then read the value again, and keep trying until we know what version to use. If we read a version
+we don't support yet, we throw (presumably, this is part of a rolling upgrade and should resolve itself soon).
+
+Installing transactions2 itself is done by a background thread that submits a transform that installs version 2 (or
+whatever specific version is included in config).
+
+### Live Migrations
+
+Transactions2 was written with some of the heaviest users of AtlasDB in mind. These are core Palantir services where
+shutdown upgrades are costly or even verboten, and we thus implemented a mechanism for performing upgrades without
+downtime. The key problem here involves reasoning about service nodes running different binary versions that are
+concurrently operating on an AtlasDB deployment.
+
+#### Legacy Versions
+
+AtlasDB is deployed as a library in client services. Versions of AtlasDB may be partitioned based on their support
+for Transactions2 into the following categories:
+
+- **(A1) I always use transactions1.** These versions don't even know that the coordination service exists, and they
+  will read and write from transactions1 regardless of what other nodes are doing.
+- **(A2) I know about the coordination service, but not about transactions2.** These versions of AtlasDB will check
+  the coordination service, but they don't know how to read from/write to transactions2, and so they will throw if
+  they need to write to transactions2, or if they need to read the commit timestamp of a value written to
+  transactions2.
+- **(A3) I know about the coordination service and about transactions2.** This is fairly self-explanatory.
+
+Client services come with a version of AtlasDB, and at some point they typically enable transactions2 by default.
+We can thus similarly partition them:
+
+- C1 and C2 are defined as client service versions using versions of AtlasDB that would be classified as A1 and A2
+  respectively.
+- C3 refers to client service versions that use an A3 version of AtlasDB, but don't enable transactions2.
+- C4 refers to client service versions that use an A3 version of AtlasDB and enable transactions2. (Transactions2
```

Comment:
if C3 and C4 run concurrently, does C3 behave like C2 i.e. throw because transactions2 is not enabled, or will it just transparently work?

I suppose what I'm getting at is, it being weird that C3 has it disabled but can still work. It feels like it's exposing/relying on internal config server semantics no? as opposed to being baked into the binary.

Reply:
C3 will transparently work. Unfortunately it exists very much for the purposes of no-downtime upgrades, in that rolling from C1 or C2 to C4 causes problems, while rolling from C1/C2 to C3 is okay, and rolling from C3 and C4 is okay.